packages/react-components/react-utilities/src/hooks/useMotionPresence.ts

```diff
@@ -0,0 +1,345 @@
+import * as React from 'react';
+import { useTimeout } from './useTimeout';
+
+/**
+ * CSS Typed Object Model
+ * @see https://drafts.css-houdini.org/css-typed-om-1/
+ * @see https://developer.mozilla.org/en-US/docs/Web/API/CSSUnitValue
+ */
+interface CSSUnitValue {
+  value: number;
+  readonly unit: string;
+}
+
+/**
+ * Style property map read only.
+ * @see https://developer.mozilla.org/en-US/docs/Web/API/StylePropertyMapReadOnly
+ */
+interface StylePropertyMapReadOnly {
+  [Symbol.iterator](): IterableIterator<[string, CSSUnitValue[]]>;
+
+  get(property: string): CSSUnitValue | undefined;
+  getAll(property: string): CSSUnitValue[];
+  has(property: string): boolean;
+  readonly size: number;
+}
+
+/**
+ * HTMLElement with styled map.
+ * @see https://developer.mozilla.org/en-US/docs/Web/API/Element/computedStyleMap
+ */
+type HTMLElementWithStyledMap<TElement extends HTMLElement = HTMLElement> = TElement & {
+  computedStyleMap(): StylePropertyMapReadOnly;
+};
+
+interface CSSWithNumber {
+  number(value: number): {
+    value: number;
+    readonly unit: string;
+  };
+}
+
+/**
+ * State for useMotionPresence hook.
+ */
+export type UseMotionPresenceState<TElement extends HTMLElement> = {
+  /**
+   * Ref to the element.
+   */
+  ref: React.RefCallback<TElement>;
```

Comment:
```suggestion
  ref: React.Ref<TElement>;
```

Is there any reason to use `RefCallback` here? isn't `Ref` enough?!

Reply:
The idea was to as specific as possible, as this is not a regular ref but rather a callback ref.

Reply:
This is part of the public API, if one day you want to change from a callback ref to an object ref you won't be able to. Meanwhile if you're more generic and use just `React.Ref` you can use both